File: 2020_kakao_blind_recruitment/lock_and_key.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

vector<vector<int> > rotateKey(vector<vector<int> >& key) {
    int rSize = (int)key.size();
    int cSize = (int)key[0].size();
    
    vector<vector<int> > rotatedKey(rSize, vector<int>(cSize));
    
    for (int r = 0; r < rSize; r++) {
        for (int c = 0; c < cSize; c++) {
            rotatedKey[c][rSize-r-1] = key[r][c];
        }
    }
    
    return rotatedKey;
}

bool solution(vector<vector<int>> key, vector<vector<int>> lock) {
    bool answer = false;
    
    int m = (int)key.size();
    int n = (int)lock.size();
    
    int zeroCount = 0;
    for (int r = 0; r < n; r++) {
        for (int c = 0; c < n; c++) {
            if (lock[r][c] == 0) zeroCount++;
        }
    }
        
    // 1. key 4방향으로 회전
    for (int rotate = 0; rotate < 4; rotate++) {
        for (int lR = -20; lR <= 20; lR++) {
            for (int lC = -20; lC <= 20; lC++) {
                bool ok = true;
                int cnt = 0;
                for (int kR = 0; kR < m; kR++) {
                    for (int kC = 0; kC < m; kC++) {
                        int nR = lR + kR;
                        int nC = lC + kC;
                        if (nR < 0 || nR >= n || nC < 0 || nC >= n) {
                            continue;
                        }
                        if (lock[nR][nC] == 1 && key[kR][kC] == 1) {
                            ok = false;
                            break;
                        }
                        if (lock[nR][nC] == 0 && key[kR][kC] == 1) {
                            cnt++;
                        }
                    }
                    if (!ok) break;
                }
                if (ok && zeroCount == cnt) {
                    answer = ok;
                    break;
                }
            }
            if (answer) break;
        }
        if (answer) break;
        key = rotateKey(key);
    }
    
    return answer;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    
    vector<vector<int> > key = {
        {0, 0, 0},
        {1, 0, 0},
        {0, 1, 1}
    };
    
    vector<vector<int> > lock = {
        {1, 1, 1},
        {1, 1, 0},
        {1, 0, 1}
    };
    
    cout << solution(key, lock) << '\n';
    
    return 0;
}

File: 2020_kakao_blind_recruitment/lock_and_key.js
function rotated(key) {
  const keySize = key.length;
  const rotatedKey = Array.from(Array(keySize), () => Array());

  for (let r = 0; r < keySize; r++) {
    for (let c = 0; c < keySize; c++) {
      rotatedKey[c][keySize - 1 - r] = key[r][c];
    }
  }

  return rotatedKey;
}

function isUnlock(key, lock, r, c) {
  const keySize = key.length;
  const lockSize = lock.length;
  const boardSize = lockSize + keySize * 2 - 2;
  const board = Array.from(Array(boardSize), () => Array());

  for (let bR = 0; bR < board.length; bR++) {
    for (let bC = 0; bC < board.length; bC++) {
      const lR = bR - (keySize - 1);
      const lC = bC - (keySize - 1);

      if (0 <= lR && lR < lockSize && 0 <= lC && lC < lockSize) {
        board[bR][bC] = lock[lR][lC];
        continue;
      }

      board[bR][bC] = 0;
    }
  }

  for (let bR = r; bR < r + keySize; bR++) {
    for (let bC = c; bC < c + keySize; bC++) {
      board[bR][bC] += key[bR - r][bC - c];
    }
  }

  for (let lR = keySize - 1; lR <= boardSize - keySize; lR++) {
    for (let lC = keySize - 1; lC <= boardSize - keySize; lC++) {
      if (board[lR][lC] !== 1) return false;
    }
  }

  return true;
}

function solution(key, lock) {
  const keySize = key.length;
  const lockSize = lock.length;
  const boardSize = lockSize + keySize * 2 - 2;

  for (let rotate = 0; rotate < 4; rotate++) {
    for (let r = 0; r <= boardSize - keySize; r++) {
      for (let c = 0; c <= boardSize - keySize; c++) {
        if (isUnlock(key, lock, r, c)) return true;
      }
    }
    key = rotated(key);
  }

  return false;
}

const key = [
  [0, 0, 0],
  [1, 0, 0],
  [0, 1, 1],
];

const lock = [
  [1, 1, 1],
  [1, 1, 0],
  [1, 0, 1],
];

console.log(solution(key, lock));
